hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+      }
+      return true;
+    }
+  }
+#endif
+
+  return false;
+}
+
+/*
+ * Returns true if the source is a PC Relative instruction.
+ *   ADR/ADRP
+ *   LDR (literal)
+ *   B.<cc> (immediate) - Not yet supported as it's not used.
+ *   B (immediate)
+ */
+bool relocatePCRelativeHelper(CodeBlock& dest_block, TCA start, TCA end,
+                              TCA src, TCA dest, TCA jmp_dest,
+                              size_t& src_count, size_t& dest_count,
+                              InstrSet& far, InstrSet& rewrites) {
+
+  Instruction* instr = Instruction::Cast(src);
+  Instruction* instr2 = Instruction::Cast(dest);
+
+  if (instr->IsPCRelAddressing() ||
+      instr->IsLoadLiteral() ||
+      instr->IsCondBranchImm() ||
+      instr->IsUncondBranchImm()) {
+
+    TCA target = (TCA)instr->ImmPCOffsetTarget();
+    size_t range = end - start;
+
+    // If the target is outside of the range of this relocation,
+    // then update it.
+    if (size_t(target - start) >= range) {
+      /*
+       * Calculate the new offset and determine if it can be encoded
+       * in a PC relative instruciton or if it needs to be converted
+       * to make use of an absolute target.
+       * Note: Use the VIXL scratch registers when transforming. Their
+       *       scope is just a single macroassembler directive, whereas
+       *       the scope of rAsm is an entire vasm instruction.
+       */
+      int imm = instr->ImmPCOffsetTarget() - instr2;
+      bool is_relative = true;
+      if (instr->IsPCRelAddressing()) {
+        if (!is_int21(imm) || far.count(instr)) {
+          dest_block.setFrontier(dest);
+          dest_count--;
+
+          vixl::MacroAssembler a { dest_block };
+          auto const dst = vixl::Register(instr->Rd(), 64);
+          a.Mov(dst, instr->ImmPCOffsetTarget());
+
+          dest_count += (dest_block.frontier() - dest) / kInstructionSize;
+          is_relative = false;
+        }
+      } else if (instr->IsLoadLiteral()) {
+        if (!is_int19(imm) || far.count(instr)) {
+          dest_block.setFrontier(dest);
+          dest_count--;
+
+          vixl::MacroAssembler a { dest_block };
+          a.SetScratchRegisters(vixl::NoReg, vixl::NoReg);
+          auto const dst = vixl::Register(instr->Rd(), 64);
+          auto const tmp = dst.Is(rVixlScratch0)
+            ? rVixlScratch1 : rVixlScratch0;
+          a.Mov(tmp, instr->ImmPCOffsetTarget());
+          a.Ldr(dst, vixl::MemOperand(tmp));
+          a.SetScratchRegisters(rVixlScratch0, rVixlScratch1);
+
+          dest_count += (dest_block.frontier() - dest) / kInstructionSize;
+          is_relative = false;
+        }
+      } else if (instr->IsCondBranchImm()) {
+        imm >>= vixl::kInstructionSizeLog2;
+        // TODO(CDE): Suppport relative-to-absolute relocation for
+        //            conditional branches.
+        assert_flog(is_int19(imm),
+                    "Can't yet relocate B.<cc>, imm won't fit in 19 bits.\n");
```

Comment:
I'm a little skeptical of committing code that ignores a case just because our tests don't trigger it.  You could try writing a C++ unit test (there should be some in runtime/vm/jit/tests or runtime/tests, etc.) that creates a massive Vunit and then runs relocation on it.